server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java

```diff
@@ -194,12 +198,8 @@ Set<String> sourcePaths(String field) {
         }
 
         String resolvedField = field;
-        int lastDotIndex = field.lastIndexOf('.');
-        if (lastDotIndex > 0) {
-            String parentField = field.substring(0, lastDotIndex);
-            if (fullNameToFieldType.containsKey(parentField)) {
-                resolvedField = parentField;
-            }
+        if (fullSubfieldNameToParentPath.containsKey(field)) {
+            resolvedField = fullSubfieldNameToParentPath.get(field);
         }
```

Comment:
Sorry if I missed what you meant, please try to let me know where I missed.
I'll get back to this next week.
Thanks for the review! 🙏

Reply:
Thank you for your response and clarification. I understand your concerns about the proposed solution and the performance impact of sorting the keys. At the end I proposed `Stream.max()` instead of sorting that is better but not optimal. 

Considering the limitations you've mentioned, I do agree with you, relying solely on the `fullNameToFieldType.keySet()` it’s not sufficient for correctly identifying real source paths and the use of a prefix-based approach, as you pointed out, may lead to incorrect matches.

After thinking about it some more I do agree with the [solution proposed](https://github.com/elastic/elasticsearch/pull/97092#discussion_r1254296395) by @romseygeek. 
I apologise for any confusion caused by my previous suggestion.